### PR TITLE
ListField & ManyRelatedViewModelField

### PR DESCRIPTION
**Pull request by davecoates**
As part of doing this I've found various edge cases where the caching fails to do what you would expect in regards to listeners.

I've added failing test cases for these that are currently skipped. I spent many hours trying to resolve it but it's going to take a fair bit to get there I think... probably another rewrite of the internals of caching as it's really difficult to follow.

These issues exist with or without this branch however so think we may as well proceed with this as is. This extends the existing behaviour to support RelatedViewModelField's that have multiple values instead of a single value.

### CODE REVIEW

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts [hunk 1]
```diff
@@ -167,7 +168,11 @@ function getFieldNameCacheKey(
         } else {
             // getField guarantees that the field will be a RelatedViewModelField so we
             // don't need to check it here
-            const relatedField = viewModel.getField(path.slice(0, -1)) as RelatedViewModelField;
+            const relatedField = viewModel.getField(path.slice(0, -1)) as BaseRelatedViewModelField<
+                any,
+                any,
+                any
+            >;
```
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:175]: Does `any` here really mean any, or does it mean "any, but as long as it satisfies the 'extends' constraints in the class definition"?

File: js-packages/@prestojs/viewmodel/src/fields/Field.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ import { AsyncChoicesInterface } from './AsyncChoices';
 /**
  * @expand-properties
  */
-export interface FieldProps<T> {
+export interface FieldProps<T, SingleType = T> {
```
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/Field.ts]: Can you add a comment for what T and SingleType are -- it's not clear at first glance

(I assume T is the value in the field which in the case of multivalue fields would normally be an array of SingleType?)

File: js-packages/@prestojs/viewmodel/src/fields/Field.ts [hunk 3]
```diff
@@ -67,7 +67,7 @@ class UnboundFieldError<T, K> extends Error {
  * @extract-docs
  * @menu-group Fields
  */
-export default class Field<T, ParsableType extends any = T> {
+export default class Field<T, ParsableType extends any = T, SingleType = T> {
```
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/Field.ts]: Is it documented anywhere what ParsableType is?

File: js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts [hunk 4]
```diff
@@ -1,11 +1,31 @@
...
+type RelatedViewModelParsableType<T extends ViewModelConstructor<any>> =
+    | FieldDataMappingRaw<T['fields']>
+    | FieldDataMappingRaw<T['fields']>[];
 
 /**
  * @expand-properties
  */
-type RelatedViewModelFieldProps<T extends ViewModelConstructor<any>> = FieldProps<
-    ViewModelInterface<T['fields'], any, T['__pkFieldType'], T['__pkType']>
+type RelatedViewModelFieldProps<T extends ViewModelConstructor<any>, FieldValueType> = FieldProps<
```
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts]: As a general comment, am wondering about being more explicit with types. `T` is helpful because you immediately know that it's a generic type, but it's not always obvious what the type represents.

Does ending them with T (to distinguish from regular types) make sense?

eg in this case it would be something like
`type RelatedViewModelFieldProps<TargetViewModelT extends ViewModelConstructor<any>, FieldValueT>`

File: js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts [hunk 5]
```diff
@@ -20,15 +40,148 @@ type RelatedViewModelFieldProps<T extends ViewModelConstructor<any>> = FieldProp
...
+            this._resolvedTo = to;
+        } else {
+            this._loadTo = to;
+        }
+        this.sourceFieldName = sourceFieldName;
+        this._isResolvingDeps = false;
+    }
+
+    /**
+     * @private
```
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts:94]: Should document what this function does -- as with comment on other review, it will make sense to devs familiar with django internals but not necessarily other frameworks/languages

**Review by levic — COMMENTED** [review]
Happy for this to be committed without further review (but can you @ me so I can see the answers to my questions)

**Comment by levic** [comment]
I don't see any issue for refactoring the caching code; can you add an item to the board for this.

Be reasonably specific about what you think needs to be updated -- the test cases are pretty comprehensive but if you're not around I think someone looking at it will have no idea where to start at all.

File: js-packages/@prestojs/viewmodel/src/fields/Field.ts (hunk 2 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/Field.ts]: Yep that's right

File: js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts (hunk 4 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts]: Was trying to find some existing conventions in community, best I came across was this https://docs.microsoft.com/en-gb/archive/blogs/kcwalina/design-guidelines-update-names-of-generic-type-parameters which was referenced in a discussion about typescript. Seems as good as any 🤷‍♂️

File: js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts (hunk 5 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts:94]: It's documented on the base class

File: js-packages/@prestojs/viewmodel/src/fields/Field.ts (hunk 3 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/Field.ts]: No generally haven't documented generic types, will add something

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts (hunk 1 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:175]: pretty sure it results in all constraints being ignored. i don't think it really matters here though as don't know anything specific... i guess could specify the constraints again 🤷‍♂️

File: js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts (hunk 4 above)
Comment (levic) [reply js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts]: Using T as a prefix seems a very Microsoft thing to do (eg everything in COM is prefixed with I for Interface, ), but suffixes seem more common in JS-land (but my exposure to other typescript codebases is not very strong so I may be wrong)
